syncstorage/src/web/extractors.rs

```diff
@@ -419,49 +419,50 @@ impl FromRequest for BsoBody {
-            .map_err(|e| {
-                warn!("⚠️ Could not parse BSO Body: {:?}", e);
-                let err: ApiError = ValidationErrorKind::FromDetails(
-                    e.to_string(),
-                    RequestErrorLocation::Body,
-                    Some("bso".to_owned()),
-                    label!("request.validate.bad_bso_body"),
-                )
-                .into();
-                err.into()
-            })
-            .and_then(move |bso: Json<BsoBody>| {
-                // Check the max payload size manually with our desired limit
-                if bso
-                    .payload
-                    .as_ref()
-                    .map(std::string::String::len)
-                    .unwrap_or_default()
-                    > max_payload_size
-                {
+        // We declare fut here to avoid moving a mutable ref of payload into the async block.
+        // See examples in actix documentation for this style of implementation.
+        let fut = <Json<BsoBody>>::from_request(req, payload);
+
+        Box::pin(async move {
+            let bso = fut
+                .map_err(|e| {
+                    warn!("⚠️ Could not parse BSO Body: {:?}", e);
                     let err: ApiError = ValidationErrorKind::FromDetails(
-                        "payload too large".to_owned(),
+                        e.to_string(),
                         RequestErrorLocation::Body,
                         Some("bso".to_owned()),
-                        label!("request.validate.payload_too_large"),
+                        label!("request.validate.bad_bso_body"),
                     )
                     .into();
-                    return future::err(err.into());
-                }
-                if let Err(e) = bso.validate() {
-                    let err: ApiError = ValidationErrorKind::FromValidationErrors(
-                        e,
-                        RequestErrorLocation::Body,
-                        None,
-                    )
-                    .into();
-                    return future::err(err.into());
-                }
-                future::ok(bso.into_inner())
-            });
-
-        Box::pin(fut)
+                    err
+                })
+                .await?;
+            // .and_then(move |bso: Json<BsoBody>| {
+            // Check the max payload size manually with our desired limit
+            if bso
+                .payload
+                .as_ref()
+                .map(std::string::String::len)
+                .unwrap_or_default()
+                > max_payload_size
+            {
+                let err: ApiError = ValidationErrorKind::FromDetails(
+                    "payload too large".to_owned(),
+                    RequestErrorLocation::Body,
+                    Some("bso".to_owned()),
+                    label!("request.validate.payload_too_large"),
+                )
+                .into();
+                return Err(err.into());
```

Comment:
Like above I think we can simplify these 2 while you're here (but they still need `.into()` because we don't use `?` here:
```suggestion
                return Err(ValidationErrorKind::FromDetails(
                    "payload too large".to_owned(),
                    RequestErrorLocation::Body,
                    Some("bso".to_owned()),
                    label!("request.validate.payload_too_large"),
                ).into());
```

Reply:
This is an excellent point and solid refactor, thank you for catching that. Will resolve.